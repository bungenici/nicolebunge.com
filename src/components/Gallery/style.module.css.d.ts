/* eslint-disable camelcase */

export const gallery: string;
export const gallery__item: string;
