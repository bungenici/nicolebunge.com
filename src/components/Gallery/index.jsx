/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/no-array-index-key */
import Img from 'gatsby-image';
import PropTypes from 'prop-types';
import React, { Component, Fragment } from 'react';
import Lightbox from 'react-images';
import styles from './style.module.css';

class Gallery extends Component {
  constructor(props) {
    super(props);

    this.state = {
      lightboxIsOpen: false,
      currentImage: 0,
    };
  }

  openLightbox(index, event) {
    event.preventDefault();
    this.setState({
      currentImage: index,
      lightboxIsOpen: true,
    });
  }

  closeLightbox() {
    this.setState({
      currentImage: 0,
      lightboxIsOpen: false,
    });
  }

  gotoPrevious() {
    const { currentImage } = this.state;

    this.setState({
      currentImage: currentImage - 1,
    });
  }

  gotoNext() {
    const { currentImage } = this.state;

    this.setState({
      currentImage: currentImage + 1,
    });
  }

  gotoImage(index) {
    this.setState({
      currentImage: index,
    });
  }

  handleClickImage() {
    const { currentImage } = this.state;
    const { images } = this.props;

    if (currentImage === images.length - 1) {
      return;
    }

    this.gotoNext();
  }

  render() {
    const { currentImage, lightboxIsOpen } = this.state;
    const { images } = this.props;

    return (
      <Fragment>
        <ul className={styles.gallery}>
          {images.map((image, index) => (
            <li key={index} className={styles.gallery__item}>
              <figure>
                <a href={image.fluid.srcSet} onClick={(event) => this.openLightbox(index, event)}>
                  <Img fluid={image.fluid} />
                </a>
              </figure>
            </li>
          ))}
        </ul>
        <Lightbox
          currentImage={currentImage}
          images={images.map((image) => image.fluid)}
          isOpen={lightboxIsOpen}
          onClickImage={() => this.handleClickImage()}
          onClickNext={() => this.gotoNext()}
          onClickPrev={() => this.gotoPrevious()}
          onClickThumbnail={() => this.gotoImage()}
          onClose={() => this.closeLightbox()}
        />
      </Fragment>
    );
  }
}

Gallery.propTypes = {
  images: PropTypes.arrayOf(PropTypes.shape()),
};

Gallery.defaultProps = {
  images: [],
};

export default Gallery;
