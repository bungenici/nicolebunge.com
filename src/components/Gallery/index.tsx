/* eslint-disable react/no-array-index-key */

import classNames from 'classnames';
import Img, { FluidObject } from 'gatsby-image';
import React, { Component } from 'react';
import Carousel, { Modal, ModalGateway } from 'react-images';
import styles from './style.module.css';

interface Image {
  fluid: FluidObject;
}

export interface GalleryProps {
  className?: string;
  images: Image[];
  isLoading?: boolean;
}

export interface GalleryState {
  lightboxIsOpen: boolean;
  selectedIndex: number;
}

class Gallery extends Component<GalleryProps, GalleryState> {
  constructor(props: GalleryProps) {
    super(props);

    this.state = {
      selectedIndex: 0,
      lightboxIsOpen: false,
    };
  }

  toggleLightbox = (selectedIndex: number): void => {
    this.setState((state) => ({
      lightboxIsOpen: !state.lightboxIsOpen,
      selectedIndex,
    }));
  };

  render(): JSX.Element {
    const { lightboxIsOpen, selectedIndex } = this.state;
    const { className, images, isLoading, ...otherProps } = this.props;
    const mappedImages = images.map((image) => ({ source: image.fluid.src }));

    return (
      <>
        <ul className={classNames(className, styles.gallery)} {...otherProps}>
          {images.map((image, index) => (
            <li key={index} className={styles.gallery__item}>
              <figure>
                <button onClick={(): void => this.toggleLightbox(index)} type="button">
                  <Img fluid={image.fluid} />
                </button>
              </figure>
            </li>
          ))}
        </ul>
        <ModalGateway>
          {lightboxIsOpen && !isLoading ? (
            <Modal onClose={this.toggleLightbox}>
              <Carousel
                currentIndex={selectedIndex}
                frameProps={{ autoSize: 'height' }}
                views={mappedImages}
              />
            </Modal>
          ) : null}
        </ModalGateway>
      </>
    );
  }
}

export default Gallery;
