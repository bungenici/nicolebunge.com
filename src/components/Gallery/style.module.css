.gallery {
  display: grid;
  grid-gap: 1.5em;
  grid-template-columns: repeat(3, 1fr);
  list-style: none;
  margin: 0;
  padding: 0;
}

.gallery__item:first-child {
  grid-row-end: span 2;
  grid-row-start: 1;
}

.gallery__figure {
  height: 100%;
}

.gallery__image {
  height: 100%;
  object-fit: cover;
  width: 100%;
}
