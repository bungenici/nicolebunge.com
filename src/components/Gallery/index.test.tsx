/* eslint-disable import/no-extraneous-dependencies */

import { render } from '@testing-library/react';
import React from 'react';
import Gallery from '.';

describe('Gallery component', () => {
  it('renders correctly', () => {
    const images = [
      {
        fluid: {
          aspectRatio: 1,
          sizes: '',
          src: '',
          srcSet: '',
        },
      },
    ];
    const { asFragment } = render(<Gallery images={images} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
