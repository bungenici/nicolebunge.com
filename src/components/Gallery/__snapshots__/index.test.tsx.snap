// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery component renders correctly 1`] = `
<ul
  className="gallery"
>
  <li
    className="gallery__item"
  >
    <figure>
      <button
        onClick={[Function]}
        type="button"
      >
        <div
          className=" gatsby-image-wrapper"
          style={
            Object {
              "maxHeight": null,
              "maxWidth": null,
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            aria-hidden={true}
            style={
              Object {
                "paddingBottom": "100%",
                "width": "100%",
              }
            }
          />
          <noscript
            dangerouslySetInnerHTML={
              Object {
                "__html": "<picture><source srcset=\\"\\" /><img loading=\\"lazy\\" src=\\"\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
              }
            }
          />
        </div>
      </button>
    </figure>
  </li>
</ul>
`;
